### ISSUE TITLE
Thoughts about `DocumentStart` having the incorrect marker/span

### ISSUE THREAD

**Issue by felipesere** [saphyr-rs/saphyr-parser#8]
👋 I've been trying to dig into markser/spans and how they are produced.

To begin with, I was looking into why `DocumentStart` was getting getting the wrong marker for a simple 
`foo: bar` document:

```
Parser state: StreamStart
    ↳ StreamStart(Utf8) Span { start: Marker { index: 0, line: 1, col: 0 }, end: Marker { index: 0, line: 1, col: 0 } }
      ↳ StreamStart

Parser state: ImplicitDocumentStart
  → fetch_next_token after whitespace Marker { index: 0, line: 1, col: 0 } 'f'
  → fetch_next_token after whitespace Marker { index: 3, line: 1, col: 3 } ':'
    ↳ BlockMappingStart Span { start: Marker { index: 3, line: 1, col: 3 }, end: Marker { index: 3, line: 1, col: 3 } }
      ↳ DocumentStart(false)

Parser state: BlockNode
      ↳ MappingStart(0, None)

Parser state: BlockMappingFirstKey
    ↳ Key Span { start: Marker { index: 0, line: 1, col: 0 }, end: Marker { index: 0, line: 1, col: 0 } }
    ↳ Scalar(Plain, "foo") Span { start: Marker { index: 0, line: 1, col: 0 }, end: Marker { index: 3, line: 1, col: 3 } }
      ↳ Scalar("foo", Plain, 0, None)
```

I dded a bunch more printlns/dbgs! to understand how the code flows and where the values come from.
From that I have somewhat narrowed it down to:

```rust
fn fetch_value(&mut self) -> ScanResult {
...
  let start_mark = self.mark;
  ...
			// Add the BLOCK-MAPPING-START token if needed.
      self.roll_indent(
         sk.mark.col,
         Some(sk.token_number),
         TokenType::BlockMappingStart,
         start_mark,
      );
  ...
}
```

`self.mark` is at this point has already advanced to `(3,1,3)` as we have processed `Scalar(Plain, "foo")`, which can be found pushed to the back of `self.tokens`.
I don't really know what the "correct" thing to do is, but intuitively feel like rather than using `self.marker` for some of these tokens - the ones that don't really consume any characters from the input - is to backtrack through `self.tokens` and find a more reasonable marker. 

Basically distinguishing between Tokens which actually consume characters and those that don't?

**Comment by Ethiraric** [comment]
I don't even think dividing the tokens into 2 categories would work. 

There are some times where we need to do some large lookahead before emitting tokens. I think namely of scalars (where there can be spaces at the end of the line) and sequences (where some characters are not allowed after the `-`. 

I have gone over all the scanning code in the last year. I probably couldn't list all instances of this. I do not think I can even guess remotely close to how hard the task is overall.

The good thing is that it's fairly easy to unit test that.

**Comment by Ethiraric** [comment]
Closing as there has been contributions towards that goal and this probably is now stale.
